File: backend/app/db/session/retry.py
# -*- coding: utf-8 -*-
"""
Retry logic for database operations.
"""

from time import sleep
from typing import Callable, TypeVar

from loguru import logger

# ##: Type variable for generic return types.
T = TypeVar("T")


def with_retry(max_retries: int, retry_delay: float):
    """
    Decorator for retrying database operations on failure.

    Parameters
    ----------
    max_retries : int
        Maximum number of retry attempts
    retry_delay : float
        Delay between retries in seconds

    Example:
    --------
    ```python
    @with_retry(max_retries=5, retry_delay=1.0)
    def get_player(player_id):
        # Function implementation
    ```
    """

    def decorator(func: Callable[..., T]) -> Callable[..., T]:
        def wrapper(*args, **kwargs) -> T:
            last_exception = None
            for attempt in range(max_retries):
                try:
                    return func(*args, **kwargs)
                except Exception as exc:
                    last_exception = exc
                    logger.warning(f"Attempt {attempt + 1}/{max_retries} failed: {exc}. Retrying in {retry_delay}s...")
                    if attempt < max_retries - 1:
                        sleep(retry_delay)

            logger.error(f"All {max_retries} attempts failed. Last error: {last_exception}")
            raise last_exception

        return wrapper

    return decorator

File: lib/db/retry.ts
// [>]: Retry utility for database operations. Matches Python backend pattern.

export interface RetryOptions {
  maxRetries?: number;
  retryDelay?: number;
}

const DEFAULT_MAX_RETRIES = 3;
const DEFAULT_RETRY_DELAY = 500;

type AsyncFunction<T extends unknown[], R> = (...args: T) => Promise<R>;

/**
 * Wraps an async function with retry logic.
 *
 * @param fn - The async function to wrap.
 * @param options - Configuration for max retries and delay between attempts.
 * @returns A wrapped function that retries on failure.
 */
export function withRetry<T extends unknown[], R>(
  fn: AsyncFunction<T, R>,
  options: RetryOptions = {},
): AsyncFunction<T, R> {
  const { maxRetries = DEFAULT_MAX_RETRIES, retryDelay = DEFAULT_RETRY_DELAY } =
    options;

  return async (...args: T): Promise<R> => {
    // [>]: Preserve original error type for instanceof checks in error handlers.
    let lastError: unknown = null;

    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        return await fn(...args);
      } catch (error) {
        lastError = error;
        const message = error instanceof Error ? error.message : String(error);
        console.warn(`Attempt ${attempt}/${maxRetries} failed: ${message}`);

        if (attempt < maxRetries) {
          await new Promise((resolve) => setTimeout(resolve, retryDelay));
        }
      }
    }

    const message =
      lastError instanceof Error ? lastError.message : String(lastError);
    console.error(`All ${maxRetries} attempts failed. Last error: ${message}`);
    throw lastError;
  };
}
